workflow/rules/preprocess.smk

```diff
@@ -12,17 +12,104 @@ phenotype_samples_df = pd.read_csv(PHENOTYPE_SAMPLES_FP, sep="\t")
 print(f"SBS samples shape: {sbs_samples_df.shape}")
 print(f"Phenotype samples shape: {phenotype_samples_df.shape}")
 
+
 # Extract metadata for SBS images
 rule extract_metadata_sbs:
+    conda:
+        "../envs/preprocess.yml"
+    input:
+        lambda wildcards: get_sample_fps(
+            sbs_samples_df, well=wildcards.well, cycle=wildcards.cycle
```

Comment:
I haven't seen this before, but get_sample_fps is a function that returns wildcards?